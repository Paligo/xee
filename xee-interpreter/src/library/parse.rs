// https://www.w3.org/TR/xpath-functions-31/#parsing-and-serializing

use xee_xpath_macros::xpath_fn;

use crate::{context, error, interpreter::Interpreter, sequence, wrap_xpath_fn};

use super::StaticFunctionDescription;

#[xpath_fn("fn:parse-xml($arg as xs:string?) as document-node(element(*))?")]
fn parse_xml(
    context: &context::DynamicContext,
    interpreter: &mut Interpreter,
    arg: Option<&str>,
) -> error::Result<Option<xot::Node>> {
    if let Some(arg) = arg {
        let documents = context.documents();
        let handle = documents
            .borrow_mut()
            .add_string(interpreter.xot_mut(), None, arg)
            .map_err(|_| error::Error::FODC0006)?;
        let doc = documents
            .borrow()
            .get_node_by_handle(handle)
            .ok_or(error::Error::FODC0006)?;
        Ok(Some(doc))
    } else {
        Ok(None)
    }
}

#[xpath_fn("fn:parse-xml-fragment($arg as xs:string?) as document-node()?")]
fn parse_xml_fragment(
    context: &context::DynamicContext,
    interpreter: &mut Interpreter,
    arg: Option<&str>,
) -> error::Result<Option<xot::Node>> {
    if let Some(arg) = arg {
        let documents = context.documents();
        let handle = documents
            .borrow_mut()
            .add_fragment_string(interpreter.xot_mut(), arg)
            .map_err(|_| error::Error::FODC0006)?;
        let doc = documents
            .borrow()
            .get_node_by_handle(handle)
            .ok_or(error::Error::FODC0006)?;
        Ok(Some(doc))
    } else {
        Ok(None)
    }
}

#[xpath_fn("fn:serialize($arg as item()*) as xs:string")]
fn serialize1(interpreter: &mut Interpreter, arg: &sequence::Sequence) -> error::Result<String> {
    let node = arg.normalize(" ", interpreter.xot_mut())?;
    Ok(interpreter.xot().to_string(node)?)
}

#[xpath_fn("fn:serialize($arg as item()*, $params as item()?) as xs:string")]
fn serialize2(
    interpreter: &mut Interpreter,
    arg: &sequence::Sequence,
    _params: Option<sequence::Item>,
) -> error::Result<String> {
    // TODO: nothing is done with the parameters yet
    let node = arg.normalize(" ", interpreter.xot_mut())?;
    Ok(interpreter.xot().to_string(node)?)
}

pub(crate) fn static_function_descriptions() -> Vec<StaticFunctionDescription> {
    vec![
        wrap_xpath_fn!(parse_xml),
        wrap_xpath_fn!(parse_xml_fragment),
        wrap_xpath_fn!(serialize1),
        wrap_xpath_fn!(serialize2),
    ]
}
