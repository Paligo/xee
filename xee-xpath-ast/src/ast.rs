mod ast_core;
mod parse_ast;
mod rename;
mod visitor;

pub use ast_core::*;
pub use parse_ast::{parse_expr_single, parse_xpath};
