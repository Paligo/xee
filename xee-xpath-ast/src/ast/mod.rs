mod ast_core;
mod parser;
mod rename;
mod visitor;

pub use ast_core::*;
pub use parser::{
    parse_expr_single, parse_kind_test, parse_sequence_type, parse_signature, parse_xpath,
};
