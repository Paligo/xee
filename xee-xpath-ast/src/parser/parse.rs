// parse XPath via pest

// XXX extra-grammatical constraints still need to be handled:
// https://www.w3.org/TR/xpath-31/#extra-grammatical-constraints
// XXX reserved function names still need to be handled:
// https://www.w3.org/TR/xpath-31/#id-reserved-fn-names

#[derive(Parser)]
#[grammar = "parser/xpath-31.pest"]
pub(crate) struct XPathParser;

#[cfg(test)]
mod tests {
    use super::*;
    use pest::iterators::Pairs;
    use pest::Parser;

    // parse function signature as described in
    // https://www.w3.org/TR/xpath-functions-31/#func-signatures
    // This is almost the same as an inline function definition except
    // for the name part.
    #[allow(clippy::result_large_err)]
    pub(crate) fn parse_signature(
        signature: &str,
    ) -> Result<Pairs<Rule>, pest::error::Error<Rule>> {
        XPathParser::parse(Rule::Signature, signature)
    }

    #[test]
    fn test_char() {
        let successful_parse = XPathParser::parse(Rule::Char, "abc");
        assert!(successful_parse.is_ok());
        let unsuccessful_parse = XPathParser::parse(Rule::Char, "\u{10}");
        assert!(unsuccessful_parse.is_err());
    }

    #[test]
    fn test_not_xpath() {
        let unsuccessful_parse = XPathParser::parse(Rule::Xpath, ")(");
        assert!(unsuccessful_parse.is_err());
    }

    #[test]
    fn test_not_xpath_incomplete() {
        let unsuccessful_parse = XPathParser::parse(Rule::Xpath, "foo(");
        assert!(unsuccessful_parse.is_err());
    }

    #[test]
    fn test_xpath_simple_name() {
        let successful_parse = XPathParser::parse(Rule::Xpath, "foo");
        assert!(successful_parse.is_ok());
    }

    #[test]
    fn test_xpath_relative_path() {
        let successful_parse = XPathParser::parse(Rule::Xpath, "foo/bar");
        assert!(successful_parse.is_ok());
    }

    #[test]
    fn test_xpath_absolute_path() {
        let successful_parse = XPathParser::parse(Rule::Xpath, "/foo/bar");
        assert!(successful_parse.is_ok());
    }

    #[test]
    fn test_signature() {
        let successful_parse = parse_signature("math:exp($arg as xs:double?) as xs:double?");
        assert!(successful_parse.is_ok());
    }

    #[test]
    fn test_signature_parameter_type_required() {
        let successful_parse = parse_signature("math:exp($arg) as xs:double?");
        assert!(successful_parse.is_err());
    }

    #[test]
    fn test_signature_return_type_required() {
        let successful_parse = parse_signature("math:exp($arg as xs:double?)");
        assert!(successful_parse.is_err());
    }
}
