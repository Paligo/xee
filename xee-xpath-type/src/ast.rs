use xee_name::Name;
use xee_schema_type::Xs;

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum SequenceType {
    Empty,
    Item(Item),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct Item {
    pub item_type: ItemType,
    pub occurrence: Occurrence,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum ItemType {
    Item,
    AtomicOrUnionType(Xs),
    KindTest(KindTest),
    FunctionTest(FunctionTest),
    MapTest(MapTest),
    ArrayTest(ArrayTest),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum Occurrence {
    One,
    Option,
    Many,
    NonEmpty,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum KindTest {
    Document(Option<DocumentTest>),
    Element(Option<ElementOrAttributeTest>),
    Attribute(Option<ElementOrAttributeTest>),
    SchemaElement(SchemaElementTest),
    SchemaAttribute(SchemaAttributeTest),
    PI(Option<PITest>),
    Comment,
    Text,
    NamespaceNode,
    Any,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum DocumentTest {
    Element(Option<ElementOrAttributeTest>),
    SchemaElement(SchemaElementTest),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct ElementOrAttributeTest {
    pub name_or_wildcard: NameOrWildcard,
    pub type_name: Option<TypeName>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct TypeName {
    pub name: Xs,
    // only relevant for elements; for attributes it's always true
    pub can_be_nilled: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum NameOrWildcard {
    Name(Name),
    Wildcard,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct SchemaElementTest {
    pub name: Name,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct SchemaAttributeTest {
    pub name: Name,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum FunctionTest {
    AnyFunctionTest,
    TypedFunctionTest(Box<TypedFunctionTest>),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct TypedFunctionTest {
    pub parameter_types: Vec<SequenceType>,
    pub return_type: SequenceType,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum MapTest {
    AnyMapTest,
    TypedMapTest(Box<TypedMapTest>),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct TypedMapTest {
    pub key_type: Xs,
    pub value_type: SequenceType,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum ArrayTest {
    AnyArrayTest,
    TypedArrayTest(Box<TypedArrayTest>),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct TypedArrayTest {
    pub item_type: SequenceType,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub enum PITest {
    Name(String),
    StringLiteral(String),
}
