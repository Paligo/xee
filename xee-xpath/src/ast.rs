mod ast_core;
mod parse_ast;
mod rename;
mod visitor;

pub use ast_core::Name;
pub(crate) use ast_core::*;
pub(crate) use parse_ast::{parse_expr_single, parse_xpath};
