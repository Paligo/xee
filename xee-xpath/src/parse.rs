// parse XPath via pest

// XXX extra-grammatical constraints still need to be handled:
// https://www.w3.org/TR/xpath-31/#extra-grammatical-constraints
// XXX reserved function names still need to be handled:
// https://www.w3.org/TR/xpath-31/#id-reserved-fn-names

#[derive(Parser)]
#[grammar = "xpath-31.pest"]
pub struct XPathParser;

#[cfg(test)]
mod tests {
    use super::*;
    use pest::Parser;

    #[test]
    fn test_char() {
        let successful_parse = XPathParser::parse(Rule::Char, "abc");
        assert!(successful_parse.is_ok());
        let unsuccessful_parse = XPathParser::parse(Rule::Char, "\u{10}");
        assert!(unsuccessful_parse.is_err());
    }

    #[test]
    fn test_xpath() {
        let successful_parse = XPathParser::parse(Rule::Xpath, "/a/c");
        assert!(successful_parse.is_ok());
        let unsuccessful_parse = XPathParser::parse(Rule::Xpath, ")(");
        assert!(unsuccessful_parse.is_err());
    }
}
