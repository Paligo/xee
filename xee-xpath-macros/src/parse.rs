use syn::{
    parse::{Parse, ParseStream},
    token::Comma,
    LitStr, Result,
};

use xee_xpath_ast::ast::Signature;
use xee_xpath_ast::parse_signature;
use xee_xpath_ast::Namespaces;

#[derive(Debug)]
pub(crate) struct XPathFnOptions {
    pub(crate) signature: Signature,
    pub(crate) kind: Option<String>,
    pub(crate) signature_string: String,
}

mod kw {
    syn::custom_keyword!(context_first);
    syn::custom_keyword!(context_last);
    syn::custom_keyword!(position);
    syn::custom_keyword!(size);
}

impl Parse for XPathFnOptions {
    fn parse(input: ParseStream) -> Result<Self> {
        let mut options = Vec::new();
        while !input.is_empty() {
            options.push(input.parse()?);
            if !input.is_empty() {
                let _: Comma = input.parse()?;
            }
        }

        let mut signature: Option<String> = None;
        let mut kind: Option<String> = None;
        for option in options {
            match option {
                XPathFnOption::Signature(signature_option) => {
                    signature = Some(signature_option);
                }
                XPathFnOption::Kind(kind_option) => {
                    kind = Some(kind_option);
                }
            }
        }
        let signature_string = signature.unwrap();
        let namespaces = Namespaces::default();
        let signature = parse_signature(&signature_string, &namespaces)
            .map_err(|e| syn::Error::new(input.span(), format!("{:?}", e)))?;
        Ok(Self {
            signature,
            kind,
            signature_string,
        })
    }
}

enum XPathFnOption {
    Signature(String),
    Kind(String),
}

impl Parse for XPathFnOption {
    fn parse(input: ParseStream) -> Result<Self> {
        let lookahead = input.lookahead1();
        Ok(if lookahead.peek(kw::context_first) {
            let _eat: kw::context_first = input.parse()?;
            XPathFnOption::Kind("context_first".to_string())
        } else if lookahead.peek(kw::context_last) {
            XPathFnOption::Kind("context_last".to_string())
        } else if lookahead.peek(kw::position) {
            XPathFnOption::Kind("position".to_string())
        } else if lookahead.peek(kw::size) {
            XPathFnOption::Kind("size".to_string())
        } else if lookahead.peek(LitStr) {
            let string_literal: LitStr = input.parse()?;
            let signature = string_literal.value();
            XPathFnOption::Signature(signature)
        } else {
            bail_spanned!(
                input.span() => "Expected a string literal or a context keyword"
            );
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_parse_signature() {
        assert_debug_snapshot!(syn::parse_str::<XPathFnOptions>(
            r#""fn:foo() as xs:string""#
        ));
    }

    #[test]
    fn test_parse_signature_with_kind() {
        assert_debug_snapshot!(syn::parse_str::<XPathFnOptions>(
            r#""fn:foo() as xs:string", context_first"#
        ));
    }

    #[test]
    fn test_parse_signature_parse_error() {
        assert_debug_snapshot!(syn::parse_str::<XPathFnOptions>(r#""wrong wrong""#));
    }

    #[test]
    fn test_parse_not_a_signature_string() {
        assert_debug_snapshot!(syn::parse_str::<XPathFnOptions>(r#"wrong"#));
    }

    #[test]
    fn test_parse_signature_unknown_kind() {
        assert_debug_snapshot!(syn::parse_str::<XPathFnOptions>(
            r#""fn:foo() as xs:string",blah"#
        ));
    }
}
