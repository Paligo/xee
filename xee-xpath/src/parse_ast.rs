use ordered_float::OrderedFloat;
use pest::iterators::Pair;

use crate::ast;
use crate::parse::{parse, Rule};

pub struct Error {}

fn struct_wrap<T, W>(pair: Pair<Rule>, outer_rule: Rule, inner_rule: Rule, wrap: W) -> T
where
    W: Fn(Pair<Rule>) -> T,
{
    debug_assert_eq!(pair.as_rule(), outer_rule);
    let pair = pair.into_inner().next().unwrap();
    if pair.as_rule() == inner_rule {
        wrap(pair)
    } else {
        panic!("unhandled {:?}", pair.as_rule())
    }
}

fn pair_to_path_expr(pair: Pair<Rule>) -> ast::PathExpr {
    let expr_single = expr_single(pair);
    match expr_single {
        ast::ExprSingle::Path(path_expr) => path_expr,
        _ => ast::PathExpr {
            steps: vec![ast::StepExpr::PrimaryExpr(ast::PrimaryExpr::Expr(vec![
                expr_single,
            ]))],
        },
    }
}

fn expr_single(pair: Pair<Rule>) -> ast::ExprSingle {
    match pair.as_rule() {
        Rule::PathExpr => ast::ExprSingle::Path(path_expr_to_path_expr(pair)),
        Rule::SimpleMapExpr => {
            let mut pairs = pair.into_inner();
            let path_expr_pair = pairs.next().unwrap();
            let simple_map_path_exprs = pairs.map(pair_to_path_expr).collect::<Vec<_>>();
            if !simple_map_path_exprs.is_empty() {
                let path_expr = pair_to_path_expr(path_expr_pair);
                ast::ExprSingle::Apply(ast::ApplyExpr {
                    path_expr,
                    operator: ast::ApplyOperator::SimpleMap(simple_map_path_exprs),
                })
            } else {
                expr_single(path_expr_pair)
            }
        }
        Rule::UnaryExpr => {
            let mut plus_minus = vec![];
            for pair in pair.into_inner() {
                match pair.as_rule() {
                    Rule::Minus => {
                        plus_minus.push(ast::UnaryOperator::Minus);
                    }
                    Rule::Plus => {
                        plus_minus.push(ast::UnaryOperator::Plus);
                    }
                    Rule::ValueExpr => {
                        if plus_minus.is_empty() {
                            return expr_single(pair);
                        }
                        let path_expr = pair_to_path_expr(pair);
                        return ast::ExprSingle::Apply(ast::ApplyExpr {
                            path_expr,
                            operator: ast::ApplyOperator::Unary(plus_minus),
                        });
                    }
                    _ => {
                        panic!("unhandled unary {:?}", pair.as_rule())
                    }
                }
            }
            unreachable!();
        }
        Rule::ArrowExpr => {
            let pair = pair.into_inner().next().unwrap();
            expr_single(pair)

            // ast::ExprSingle::Path(pair_to_path_expr(pair))
        }
        Rule::CastExpr => {
            let pair = pair.into_inner().next().unwrap();
            expr_single(pair)

            // ast::ExprSingle::Path(pair_to_path_expr(pair))
        }
        Rule::CastableExpr => {
            let pair = pair.into_inner().next().unwrap();
            expr_single(pair)

            // ast::ExprSingle::Path(pair_to_path_expr(pair))
        }
        Rule::TreatExpr => {
            let pair = pair.into_inner().next().unwrap();
            expr_single(pair)

            // ast::ExprSingle::Path(pair_to_path_expr(pair))
        }
        Rule::InstanceofExpr => {
            let pair = pair.into_inner().next().unwrap();
            expr_single(pair)

            // ast::ExprSingle::Path(pair_to_path_expr(pair))
        }
        Rule::AdditiveExpr => {
            let mut pairs = pair.into_inner();
            let left_pair = pairs.next().unwrap();
            let op = pairs.next();
            if let Some(op) = op {
                let operator = match op.as_rule() {
                    Rule::Plus => ast::Operator::Add,
                    Rule::Minus => ast::Operator::Sub,
                    _ => {
                        panic!("unhandled AdditiveExpr {:?}", op.as_rule())
                    }
                };
                let right_pair = pairs.next().unwrap();
                ast::ExprSingle::Binary(ast::BinaryExpr {
                    operator,
                    left: pair_to_path_expr(left_pair),
                    right: pair_to_path_expr(right_pair),
                })
            } else {
                expr_single(left_pair)
            }
        }
        Rule::OrExpr => {
            let mut pairs = pair.into_inner();
            let left_pair = pairs.next().unwrap();
            let right_pair = pairs.next();
            if let Some(right_pair) = right_pair {
                ast::ExprSingle::Binary(ast::BinaryExpr {
                    operator: ast::Operator::Or,
                    left: pair_to_path_expr(left_pair),
                    right: pair_to_path_expr(right_pair),
                })
            } else {
                expr_single(left_pair)
            }
        }
        Rule::AndExpr
        | Rule::ComparisonExpr
        | Rule::StringConcatExpr
        | Rule::RangeExpr
        | Rule::MultiplicativeExpr
        | Rule::UnionExpr
        | Rule::IntersectExceptExpr => {
            let pair = pair.into_inner().next().unwrap();
            expr_single(pair)
            // ast::ExprSingle::Path(pair_to_path_expr(pair))
        }
        Rule::ValueExpr => {
            let pair = pair.into_inner().next().unwrap();
            expr_single(pair)
        }
        Rule::ExprSingle => {
            let pair = pair.into_inner().next().unwrap();
            expr_single(pair)
        }
        _ => {
            panic!("unhandled ExprSingle {:?}", pair.as_rule())
        }
    }
}

// fn apply_expr(pair: Pair<Rule>) -> ast::ApplyExpr {
//     match pair.as_rule() {
//         Rule::SimpleMapExpr => {}
//         Rule::UnaryExpr => {}
//         Rule::ArrowExpr => {}
//         Rule::CastExpr => {}
//         Rule::CastableExpr => {}
//         Rule::TreatExpr => {}
//         Rule::InstanceOfExpr => {}
//         _ => {
//             panic!("unhandled ApplyExpr {:?}", pair.as_rule())
//         }
//     }
// }

fn path_expr_to_path_expr(pair: Pair<Rule>) -> ast::PathExpr {
    debug_assert_eq!(pair.as_rule(), Rule::PathExpr);
    let pair = pair.into_inner().next().unwrap();
    match pair.as_rule() {
        Rule::RelativePathExpr => ast::PathExpr {
            steps: relative_path_expr_to_steps(pair),
        },
        _ => {
            panic!("unhandled PathExpr: {:?}", pair.as_rule())
        }
    }
}

fn relative_path_expr_to_steps(pair: Pair<Rule>) -> Vec<ast::StepExpr> {
    debug_assert_eq!(pair.as_rule(), Rule::RelativePathExpr);
    let pair = pair.into_inner().next().unwrap();
    match pair.as_rule() {
        Rule::StepExpr => {
            vec![step_expr_to_step_expr(pair)]
        }
        _ => {
            panic!("unhandled RelativePathExpr: {:?}", pair.as_rule())
        }
    }
}

fn step_expr_to_step_expr(pair: Pair<Rule>) -> ast::StepExpr {
    debug_assert_eq!(pair.as_rule(), Rule::StepExpr);
    let pair = pair.into_inner().next().unwrap();
    match pair.as_rule() {
        Rule::PostfixExpr => {
            let pair = pair.into_inner().next().unwrap();
            ast::StepExpr::PrimaryExpr(primary_expr_to_primary_expr(pair))
        }
        _ => {
            panic!("unhandled StepExpr: {:?}", pair.as_rule())
        }
    }
}

fn primary_expr_to_primary_expr(pair: Pair<Rule>) -> ast::PrimaryExpr {
    debug_assert_eq!(pair.as_rule(), Rule::PrimaryExpr);
    let pair = pair.into_inner().next().unwrap();
    match pair.as_rule() {
        Rule::Literal => ast::PrimaryExpr::Literal(literal_to_literal(pair)),
        _ => {
            panic!("unhandled PrimaryExpr: {:?}", pair.as_rule())
        }
    }
}

fn literal_to_literal(pair: Pair<Rule>) -> ast::Literal {
    debug_assert_eq!(pair.as_rule(), Rule::Literal);
    let pair = pair.into_inner().next().unwrap();
    match pair.as_rule() {
        Rule::StringLiteral => ast::Literal::String(pair.as_str().to_string()),
        Rule::NumericLiteral => numeric_literal_to_literal(pair),
        _ => {
            panic!("unhandled literal: {:?}", pair.as_rule())
        }
    }
}

fn numeric_literal_to_literal(pair: Pair<Rule>) -> ast::Literal {
    debug_assert_eq!(pair.as_rule(), Rule::NumericLiteral);
    let pair = pair.into_inner().next().unwrap();
    match pair.as_rule() {
        Rule::IntegerLiteral => {
            let s = pair.as_str();
            // parser never delivers negative numbers
            let i = s.parse::<i64>().unwrap();
            ast::Literal::Integer(i)
        }
        Rule::DecimalLiteral => {
            let s = pair.as_str();
            let period_index = s.find('.').unwrap();
            let (before, after) = s.split_at(period_index);
            let after = &after[1..];
            let before_nr = if !before.is_empty() {
                before.parse::<i64>().unwrap()
            } else {
                0
            };
            let after_nr = if !after.is_empty() {
                after.parse::<i64>().unwrap()
            } else {
                0
            };
            let digits = after.len();
            // to get positive number
            let factor = 10i64.pow(digits as u32);
            let before_nr = before_nr * factor;
            let nr = before_nr + after_nr;
            ast::Literal::Decimal(ast::DecimalLiteral {
                value: nr,
                fraction_digits: digits as u8,
            })
        }
        Rule::DoubleLiteral => {
            let s = pair.as_str();
            let f = s.parse::<f64>().unwrap();
            ast::Literal::Double(OrderedFloat(f))
        }
        _ => {
            panic!("unhandled numeric literal: {:?}", pair.as_rule())
        }
    }
}

// fn pair_to_ast(pair: Pair<Rule>) -> Result<ast::XPath, Error> {
//     match pair.as_rule() {
//         Rule::Xpath => {
//             let mut ast = ast::XPath { exprs: vec![] };
//             for pair in pair.into_inner() {
//                 let expr = pair_to_ast(pair)?;
//                 ast.exprs.push(expr);
//             }
//             Ok(ast)
//         }
//         _ => {}
//     }
//     Ok(ast::XPath { exprs: vec![] })
// }

// pub(crate) fn parse_ast(xpath: &str) -> Result<ast::XPath, Error> {
//     match parse(xpath) {
//         Ok(mut pairs) => {
//             let ast = pair_to_ast(pairs.next().unwrap());
//             match ast {
//                 Ok(ast) => Ok(ast),
//                 Err(e) => Err(Error {}),
//             }
//         }
//         Err(e) => Err(Error {}),
//     }
// }

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parse::XPathParser;
    use insta::assert_debug_snapshot;
    use pest::Parser;

    fn parse_rule<T, F>(rule: Rule, input: &str, f: F) -> T
    where
        F: Fn(Pair<Rule>) -> T,
    {
        let mut pairs = XPathParser::parse(rule, input).unwrap();
        let pair = pairs.next().unwrap();
        f(pair)
    }

    fn parse_expr_single(input: &str) -> ast::ExprSingle {
        parse_rule(Rule::ExprSingle, input, expr_single)
    }

    fn parse_literal(input: &str) -> ast::Literal {
        parse_rule(Rule::Literal, input, literal_to_literal)
    }

    fn parse_primary_expr(input: &str) -> ast::PrimaryExpr {
        parse_rule(Rule::PrimaryExpr, input, primary_expr_to_primary_expr)
    }

    fn parse_step_expr(input: &str) -> ast::StepExpr {
        parse_rule(Rule::StepExpr, input, step_expr_to_step_expr)
    }

    fn parse_relative_path_expr(input: &str) -> Vec<ast::StepExpr> {
        parse_rule(Rule::RelativePathExpr, input, relative_path_expr_to_steps)
    }

    fn parse_path_expr(input: &str) -> ast::PathExpr {
        parse_rule(Rule::PathExpr, input, path_expr_to_path_expr)
    }

    #[test]
    fn test_integer_literal() {
        assert_debug_snapshot!(parse_literal("1"));
    }

    #[test]
    fn test_decimal_literal() {
        assert_debug_snapshot!(parse_literal("1.5"));
    }

    #[test]
    fn test_decimal_literal_no_after() {
        assert_debug_snapshot!(parse_literal("1."));
    }

    #[test]
    fn test_decimal_literal_no_before() {
        assert_debug_snapshot!(parse_literal(".5"));
    }

    #[test]
    fn test_float_lowercase_e() {
        assert_debug_snapshot!(parse_literal("1.5e0"));
    }

    #[test]
    fn test_float_upper_e() {
        assert_debug_snapshot!(parse_literal("1.5E0"));
    }

    #[test]
    fn test_primary_expr_literal() {
        assert_debug_snapshot!(parse_primary_expr("1"));
    }

    #[test]
    fn test_step_expr() {
        assert_debug_snapshot!(parse_step_expr("1"));
    }

    #[test]
    fn test_relative_path() {
        assert_debug_snapshot!(parse_relative_path_expr("1"));
    }

    #[test]
    fn test_path_expr() {
        assert_debug_snapshot!(parse_path_expr("1"));
    }

    #[test]
    fn test_integer_expr_single() {
        assert_debug_snapshot!(parse_expr_single("1"));
    }

    #[test]
    fn test_simple_map_expr() {
        assert_debug_snapshot!(parse_expr_single("1 ! 2"));
    }

    #[test]
    fn test_unary_expr() {
        assert_debug_snapshot!(parse_expr_single("-1"));
    }

    #[test]
    fn test_additive_expr() {
        assert_debug_snapshot!(parse_expr_single("1 + 2"));
    }

    #[test]
    fn test_or_expr() {
        assert_debug_snapshot!(parse_expr_single("1 or 2"));
    }
}
