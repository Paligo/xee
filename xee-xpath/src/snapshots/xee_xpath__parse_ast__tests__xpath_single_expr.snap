---
source: xee-xpath/src/parse_ast.rs
expression: "parse_xpath_no_default_ns(\"1 + 2\")"
---
Ok(
    XPath {
        exprs: [
            (
                Binary(
                    BinaryExpr {
                        operator: Add,
                        left: PathExpr {
                            steps: [
                                PrimaryExpr(
                                    (
                                        Literal(
                                            Integer(
                                                1,
                                            ),
                                        ),
                                        0..1,
                                    ),
                                ),
                            ],
                        },
                        right: PathExpr {
                            steps: [
                                PrimaryExpr(
                                    (
                                        Literal(
                                            Integer(
                                                2,
                                            ),
                                        ),
                                        4..5,
                                    ),
                                ),
                            ],
                        },
                    },
                ),
                0..5,
            ),
        ],
    },
)
